dev/NumberBox/NumberBox.cpp

```diff
@@ -0,0 +1,312 @@
+#include "ResourceAccessor.h"
+
+using namespace std;
+
+NumberBox::NumberBox()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_NumberBox);
+    SetDefaultStyleKey(this);
+}
+
+
+void NumberBox::OnApplyTemplate()
+{
+    // Initializations - Visual Components
+    winrt::IControlProtected controlProtected = *this;
+    m_TextBox = GetTemplateChildT<winrt::TextBox>(L"InputBox", controlProtected);
+    m_SpinDown = GetTemplateChildT<winrt::Button>(L"DownSpinButton", controlProtected);
+    m_SpinUp = GetTemplateChildT<winrt::Button>(L"UpSpinButton", controlProtected);
+
+    // Initializations - Visual States
+    SetSpinButtonVisualState();
+    SetHeader();
+
+    // Initializations - Interactions
+    m_SpinDown.Click({ this, &NumberBox::OnSpinDownClick });
+    m_SpinUp.Click({ this, &NumberBox::OnSpinUpClick });
+    m_TextBox.KeyUp({ this, &NumberBox::OnNumberBoxKeyUp });
+    this->PointerWheelChanged({ this, &NumberBox::OnScroll });
+
+    // Initializations - Tools, etc.
+    Formatter = winrt::DecimalFormatter();
+    IRounder = winrt::IncrementNumberRounder();
+    SRounder = winrt::SignificantDigitsNumberRounder();
+    UpdateFormatter();
+    UpdateRounder();
+    // Set Text to reflect preset Value
+    if (s_ValueProperty != DEFAULTVALUE && m_TextBox) {
+        UpdateTextToValue();
+    }
+
+    // Register LostFocus Event
+    if ( m_TextBox ) {
+        m_TextBox.LostFocus({ this, &NumberBox::OnTextBoxLostFocus });
+    }
+}
+
+void  NumberBox::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+    
+    // Update visual state for spin buttons if placement mode changed
+    if (property == s_SpinButtonPlacementModeProperty)
+    {
+        SetSpinButtonVisualState();
+    }
+    else if (property == s_IntegerDigitsProperty || property == s_FractionDigitsProperty || property == s_SignificantDigitsProperty || property == s_IsDecimalPointAlwaysDisplayedProperty || property == s_IsZeroSignedProperty)
+    {
+        UpdateFormatter();
+    }
+    else if (property == s_RoundingAlgorithmProperty || property == s_NumberRounderProperty || property == s_SignificantDigitPrecisionProperty || property == s_IncrementPrecisionProperty)
+    {
+        UpdateRounder();
+    }
+    else if (property == s_HeaderProperty)
+    {
+        SetHeader();
+    }
+
+}
+
+// Trigger any validation, rounding, and processing done onLostFocus
+void NumberBox::OnTextBoxLostFocus(winrt::IInspectable const& sender, winrt::RoutedEventArgs const& args)
+{
+    ValidateInput();
+}
+
+// Performs all validation steps on input given in textbox. Runs on LoseFocus and stepping.
+void NumberBox::ValidateInput()
+{
+    string InputAsString = winrt::to_string(m_TextBox.Text());
+    // Handles Empty TextBox Case, current behavior is to set Value to default (0)
+    if (InputAsString == "")
+    {
+        this->Value(0);
+        HasError = false;
+        return;
```

Comment:
Should this be raising some kind of event so the markup knows that it can clear the error state?